File: Android_Zone_Test/src/com/example/mylib_test/activity/utils/Utils_MainActivity.java
package com.example.mylib_test.activity.utils;

import com.example.mylib_test.R;
import and.utils.text_edit2input.KeyBoardUtils;
import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.EditText;

public class Utils_MainActivity extends Activity implements OnClickListener{
	private EditText keyboard;
	private View view1;
	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.a_utils_test);
		keyBoardTest();
		

        DisplayMetrics metric = new DisplayMetrics();
        getWindowManager().getDefaultDisplay().getMetrics(metric);
        int width = metric.widthPixels;  // 屏幕宽度（像素）
        int height = metric.heightPixels;  // 屏幕高度（像素）
        float density = metric.density;  // 屏幕密度（0.75 / 1.0 / 1.5）
        int densityDpi = metric.densityDpi;  // 屏幕密度DPI（120 / 160 / 240）
        Log.e("Utils_MainActivity:――――――――-" + "  DisplayMetrics", "density=" + density + "; densityDPI=" + densityDpi);  
        float tem = getResources().getDimension(R.dimen.test);
        System.err.println(tem);
	
	}

	private void keyBoardTest() {
		keyboard=(EditText) findViewById(R.id.keyboard);
		view1=findViewById(R.id.view1);
		new KeyBoardUtils() {
			
			@Override
			public void openState(int keyboardHeight) {
				System.out.println("键盘：openState 高度:"+keyboardHeight);
			}
			
			@Override
			public void closeState(int keyboardHeight) {
				System.out.println("键盘：closeState 高度:"+keyboardHeight);
			}
		}.monitorKeybordState(findViewById(R.id.flowLayoutZone1));;	
	}

	@Override
	public void onClick(View v) {
		switch (v.getId()) {
		case R.id.openKeyboard:
			KeyBoardUtils.openKeybord(keyboard, this);
			break;
		case R.id.closeKeyboard:
			KeyBoardUtils.closeKeybord(keyboard, this);
			break;
		case R.id.bitmapTest:
			startActivity(new Intent(this,SoftTestActivity.class));
			break;
		case R.id.getPhone:
			startActivity(new Intent(this,GetPhoneTest.class));
			break;

		default:
			break;
		}
		
	}

}

File: Android_Zone_Test/src/com/example/mylib_test/activity/utils/Utils_MainActivity.kt
package com.example.mylib_test.activity.utils

import com.zone.lib.utils.system_hardware_software_receiver_shell.software.KeyBoardUtils

import android.content.Intent
import android.util.DisplayMetrics
import android.view.View
import com.example.mylib_test.*
import com.zone.lib.base.controller.activity.BaseFeatureActivity
import com.zone.lib.LogLevel
import com.zone.lib.ZLogger
import com.zone.lib.utils.data.file2io2data.ClipboardManagerUtils
import kotlinx.android.synthetic.main.a_utils_test.*
import com.zone.lib.base.controller.activity.controller.ActionBarActivityController
import com.zone.lib.base.controller.activity.controller.ShowState


class Utils_MainActivity : BaseFeatureActivity() {
    //https://stackoverflow.com/questions/7417123/android-how-to-adjust-layout-in-full-screen-mode-when-softkeyboard-is-visible
    //-> fullscreen mode doesn't resize  全屏模式不能 resize
    override fun initDefaultConifg(){
        getController(ActionBarActivityController::class.java)?.initFirst(ShowState.HideTitle)
    }

    override fun setContentView() {
        setContentView(R.layout.a_utils_test)
    }

    override fun initData() {
        keyBoardTest()
        val metric = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(metric)
        val width = metric.widthPixels  // 屏幕宽度（像素）
        val height = metric.heightPixels  // 屏幕高度（像素）
        val density = metric.density  // 屏幕密度（0.75 / 1.0 / 1.5）
        val densityDpi = metric.densityDpi  // 屏幕密度DPI（120 / 160 / 240）
        LogApp.i("density=$density; densityDPI=$densityDpi")
        val tem = resources.getDimension(R.dimen.test)
        System.err.println(tem)
    }

    override fun setListener() {
    }

    private fun keyBoardTest() {
        object : KeyBoardUtils() {
            override fun openState(keyboardHeight: Int) {
                LogApp.i("键盘：openState 高度:$keyboardHeight")
            }

            override fun closeState(keyboardHeight: Int) {
                LogApp.i("键盘：closeState 高度:$keyboardHeight")
            }
        }.monitorKeybordState(this@Utils_MainActivity)
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.openKeyboard -> KeyBoardUtils.openKeybord(keyboard, this)
            R.id.closeKeyboard -> KeyBoardUtils.closeKeybord(keyboard, this)
            R.id.getPhone -> startActivity(Intent(this, GetPhoneTest::class.java))
            R.id.bt_layoutClip -> startActivity(Intent(this, LayoutClipAcitivity::class.java))
            R.id.bt_log1_i -> bt_log1_i()
            R.id.bt_log12_i -> bt_log12_i()
            R.id.bt_log12_ie -> bt_log12_ie()
            R.id.getClipboardContent -> LogApp.i(ClipboardManagerUtils.getClipBoardText(this))
            R.id.setClipboardContent -> ClipboardManagerUtils.setClipeBoardContent(this,"123")
            R.id.testHide ->{
                when (testHide2.visibility) {
                    View.VISIBLE-> testHide2.visibility=View.GONE
                    View.GONE-> testHide2.visibility=View.VISIBLE
                    else -> {
                    }
                }
            }
            else -> {
            }
        }

    }

    private fun bt_log1_i() {
        ZLogger.logLevelList.clear()
        ZLogger.logLevelList.add(LogLevel.i)
        ZLogger.mayLoggerList.clear()
        ZLogger.mayLoggerList.add(Log1)
        Log1.e("eeee111")
        Log1.i("iiii111")
        Log2.e("eeee222")
        Log2.i("iiii222")

        MainActivity2.initLogger()
    }

    private fun bt_log12_i() {
        ZLogger.logLevelList.clear()
        ZLogger.logLevelList.add(LogLevel.i)
        ZLogger.mayLoggerList.clear()
        ZLogger.mayLoggerList.addAll(listOf<ZLogger>(Log1, Log2))
        Log1.e("eeee111")
        Log1.i("iiii111")
        Log2.e("eeee222")
        Log2.i("iiii222")

        MainActivity2.initLogger()
    }

    private fun bt_log12_ie() {
        ZLogger.logLevelList.clear()
        ZLogger.mayLoggerList.clear()
        ZLogger.mayLoggerList.addAll(listOf<ZLogger>(Log1, Log2))
        Log1.e("eeee111")
        Log1.i("iiii111")
        Log2.e("eeee222")
        Log2.i("iiii222")

        MainActivity2.initLogger()
    }
}
